### DISCUSSION TITLE
Use grayscale image as semi-transparent mask

### DISCUSSION THREAD

**Discussion by mviereck** [ImageMagick/ImageMagick#1971] [General] [answered]
I want to use a grayscale image as a semi-transparent mask.

My desire is to generate a mask of contrast edges. The better the contrast is, the more opaque should be the resulting pixel. Areas with no contrast should be close to transparent.

Example for a source image:
![col1 source](https://user-images.githubusercontent.com/14821372/80863184-d3785980-8c7a-11ea-953d-573c0425519b.png)
Example for a grayscale contrast mask. 
![sobelnegate](https://user-images.githubusercontent.com/14821372/80863191-db37fe00-8c7a-11ea-8320-0e7671dd33d3.png)

My attempts with different `-alpha` and `-compose` setups result in either full transparency or full opacity. Each gray pixel less than white is handled as a black one.
Can you give me a hint how to generate a semi-transparent image applying the grayscale image on the colored image, please?

**Comment by snibgo** [comment, marked as answer]
What version of IM, on what platform? Assuming IM v7:
`magick source.png mask.png -alpha off -compose CopyOpacity -composite out.png`

**Comment by fmw42** [comment]
Assuming ImageMagick 6 on Unix:

```
convert image.png \
\( +clone -colorspace gray \
-bias 50% \
-define convolve:scale='!' \
-define morphology:compose=Lighten \
-morphology Convolve 'Sobel:>' \
-contrast-stretch 0% +write image_alpha.png \) \
-alpha off -compose copy_opacity -composite \
image_with_alpha_c0.png
```

Result:

![image_with_alpha_c0](https://user-images.githubusercontent.com/13458218/80869306-9305e500-8c54-11ea-9640-0f1ee411233f.png)

Alpha (image_alpha.png)

![image_alpha](https://user-images.githubusercontent.com/13458218/80869317-9f8a3d80-8c54-11ea-8471-5821194819f7.png)

If I increase contrast to 10%:

```
convert image.png \
\( +clone -colorspace gray \
-bias 50% \
-define convolve:scale='!' \
-define morphology:compose=Lighten \
-morphology Convolve 'Sobel:>' \
-contrast-stretch 10% +write image_alpha.png \) \
-alpha off -compose copy_opacity -composite \
image_with_alpha_c10.png
```

Result:

![image_with_alpha_c10](https://user-images.githubusercontent.com/13458218/80869330-b3ce3a80-8c54-11ea-9dc8-5183d8aba6e9.png)

**Comment by mviereck** [comment]
Great, thank you, @snibgo and @fmw42 !
(Unfortunate that I can mark only one post as an answer.)

Now I can try further how to achieve usable results. The target is to generate overall sharp images from a [focus stackshot](https://en.wikipedia.org/wiki/Focus_stacking). There are some tools around already (e.g. `enfuse`), but I like to achieve this with imagemagick, too. 
One step is to find the contrasted areas of each single image, like done here.
The next step ist to combine all these contrast-masked images to one overall sharp image.

> What version of IM, on what platform? 

Sorry, you must have asked that several thousands of times ...
I have two machines (both linux/Debian) and two versions:
```
Version: ImageMagick 7.0.9-0 Q8 x86_64 2019-10-26
Version: ImageMagick 6.9.10-23 Q16 x86_64 20190101
```
(In the github options you can define a template for new issues. It makes sense to set one up for imagemagick that contains a question for the imagemagick version and platform.)

**Reply by fmw42** [reply]
For IM 7, use magick in place of convert.
